gh-132661: Disallow `Template`/`str` concatenation after PEP 750 spec update

davepeck opened the pull request:
Following the [steering council decision](https://github.com/python/steering-council/issues/289#issuecomment-2988833993) and corresponding [update to PEP750](https://github.com/python/peps/pull/4395#issuecomment-3009144017), we are removing support for both implicit and explicit `Template`/`str` concatenation.

<!-- gh-issue-number: gh-132661 -->
* Issue: gh-132661
<!-- /gh-issue-number -->

davepeck committed:
Disallow explicit concat between Template and str

davepeck committed:
First pass at removing implicit Template/str concat

sobolevn commented on Parser/action_helpers.c:
```diff
@@ -1902,6 +1907,7 @@ _PyPegen_concatenate_strings(Parser *p, asdl_expr_seq *strings,
         }
     }
 
+
```
unrelated change :)

davepeck replied:
Removed; thanks.

sobolevn commented on Objects/unicodeobject.c:
```diff
@@ -11610,16 +11609,10 @@ PyUnicode_Concat(PyObject *left, PyObject *right)
-                Py_TYPE(right)->tp_name);
-            return NULL;
-        }
+        PyErr_Format(PyExc_TypeError,
```
You can use `%T` to format object's type.

davepeck replied:
The diff here is just a dedent of existing code; I was hesitant to change anything else. But happy to update it to use `%T` if we'd prefer.

davepeck replied:
(Leaving as-is.)

sobolevn commented on Lib/test/test_tstring.py:
```diff
@@ -161,9 +161,8 @@ def test_template_concatenation(self):
-        combined = t1 + ", world"
-        self.assertTStringEqual(combined, ("Hello, world",), ())
-        self.assertEqual(fstring(combined), "Hello, world")
+        with self.assertRaises(TypeError):
```
Will it make sense to check error message here?

davepeck replied:
Good catch, will update.

sobolevn reviewed (commented):
Thank you!

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Remove extraneous newline

davepeck committed:
Add expected exception messages to tests

davepeck committed:
Remove _ast_unparse t/f implicit concat helper code

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Add a parallel comment

davepeck committed:
Add news blurb for PR.

davepeck committed:
One day, I may develop RST muscle memory.

Today was, alas, not that day.

davepeck committed:
Merge branch 'main' into pep750-concat-update

lysnikolaou commented on Lib/test/test_tstring.py:
```diff
@@ -241,45 +240,20 @@ def test_literal_concatenation(self):
+            "t'{what}-string literal' rf'raw f-string literal'",
+        ):
+            with self.assertRaisesRegex(SyntaxError, expected_msg):
+                eval(case)
```
```suggestion
            with self.subTest(case):
                with self.assertRaisesRegex(SyntaxError, expected_msg):
                    eval(case)
```

davepeck replied:
Added `self.subTest()` in the places it was missing.

lysnikolaou commented on Objects/templateobject.c:
```diff
@@ -344,46 +296,16 @@ template_concat_templates(templateobject *self, templateobject *other)
-    }
-    else {
-        Py_RETURN_NOTIMPLEMENTED;
+    } else {
```
nit

```suggestion
    }
    else {
```

sobolevn commented on Lib/test/test_tstring.py:
```diff
@@ -161,9 +160,9 @@ def test_template_concatenation(self):
-        self.assertEqual(fstring(combined), "Hello, world")
+        expected_msg = 'can only concatenate Template (not "str") to Template'
+        with self.assertRaises(TypeError, msg=expected_msg):
+            _ = t1 + ", world"
```
```suggestion
            t1 + ", world"
```

sobolevn commented on Lib/test/test_tstring.py:
```diff
@@ -174,9 +173,9 @@ def test_template_concatenation(self):
-        self.assertEqual(fstring(t), "Hello, Python")
+        expected_msg = 'can only concatenate str (not "string.templatelib.Template") to str'
+        with self.assertRaises(TypeError, msg=expected_msg):
+            _ = "Hello, " + t"{name}"
```
```suggestion
            "Hello, " + t"{name}"
```

sobolevn commented on Lib/test/test_tstring.py:
```diff
@@ -241,45 +240,20 @@ def test_literal_concatenation(self):
+            "t'{what}-string literal' u'unicode literal'",
+            "t'{what}-string literal' f'f-string literal'",
+            "t'{what}-string literal' r'raw string literal'",
+            "t'{what}-string literal' rf'raw f-string literal'",
```
Looks like we never test that `'' t'{what}'` does not work as well. Is that right?

davepeck replied:
Added tests for the other order.

sobolevn commented on Misc/NEWS.d/next/Core_and_Builtins/2025-07-08-23-22-08.gh-issue-132661.34ftJl.rst:
```diff
@@ -0,0 +1,2 @@
+Disallow concatenation of ``Template`` and ``str``. Also, disallow implicit
```
```suggestion
Reflect recent :pep:`750` change.

Disallow concatenation of :class:`string.Template` and :class:`str`. Also, disallow implicit
```

sobolevn commented on Objects/templateobject.c:
```diff
@@ -344,46 +296,16 @@ template_concat_templates(templateobject *self, templateobject *other)
+            "can only concatenate Template (not \"%.200s\") to Template",
+            Py_TYPE(other)->tp_name);
+        return NULL;
     }
```
We can simplify and dedent this:

```suggestion
    }

    PyErr_Format(PyExc_TypeError,
        "can only concatenate Template (not \"%T\") to Template",
        tp_name);
    return NULL;
```

lysnikolaou commented on Parser/action_helpers.c:
```diff
@@ -1887,6 +1887,12 @@ _PyPegen_concatenate_strings(Parser *p, asdl_expr_seq *strings,
+    if ((unicode_string_found || f_string_found) && t_string_found) {
+        RAISE_SYNTAX_ERROR("cannot mix str and Template literals");
+        return NULL;
+    }
```
This is not the way to do this. Instead, we should be doing this at the parser level. Right now we're doing this:

```
atom[expr_ty]:
    | NAME
    | 'True' { _PyAST_Constant(Py_True, NULL, EXTRA) }
    | 'False' { _PyAST_Constant(Py_False, NULL, EXTRA) }
    | 'None' { _PyAST_Constant(Py_None, NULL, EXTRA) }
    | &(STRING|FSTRING_START|TSTRING_START) strings  # This'll change
    | NUMBER
    | &'(' (tuple | group | genexp)
    | &'[' (list | listcomp)
    | &'{' (dict | set | dictcomp | setcomp)
    | '...' { _PyAST_Constant(Py_Ellipsis, NULL, EXTRA) }

strings[expr_ty] (memo): a[asdl_expr_seq*]=(fstring|string|tstring)+ { _PyPegen_concatenate_strings(p, a, EXTRA) }
```

Instead, we should be doing something like:

```
atom[expr_ty]:
    | NAME
    | 'True' { _PyAST_Constant(Py_True, NULL, EXTRA) }
    | 'False' { _PyAST_Constant(Py_False, NULL, EXTRA) }
    | 'None' { _PyAST_Constant(Py_None, NULL, EXTRA) }
    | &(STRING|FSTRING_START) strings  # We now have two alternatives
    | &TSTRING_START tstrings
    | NUMBER
    | &'(' (tuple | group | genexp)
    | &'[' (list | listcomp)
    | &'{' (dict | set | dictcomp | setcomp)
    | '...' { _PyAST_Constant(Py_Ellipsis, NULL, EXTRA) }

tstrings[expr_ty] (memo): a[asdl_expr_seq*]=tstring+ { <new action here> }
strings[expr_ty] (memo): a[asdl_expr_seq*]=(fstring|string)+ { _PyPegen_concatenate_strings(p, a, EXTRA) }
```

Notice how instead of intermixing the three rules in one sequence, we can now only mix `string`s and `fstring`s, but `tstring`s can only be concatenated with other `tstring`s. The new action should probably be `_build_concatenated_template_str` renamed with a `_PyPegen` prefix, but I'm not sure whether that'll work unchanged, it might need some editing. Also, some of the `invalid_*` rules might need changing.

davepeck replied:
Sure thing, will do.

I just wasn't sure we wanted the extra complexity in the grammar.

davepeck replied:
@lysnikolaou Would you be okay with something more along the lines of:

```
strings[expr_ty] (memo):
    | invalid_string_tstring_concat
    | a[asdl_expr_seq*]=(fstring|string)+ { _PyPegen_concatenate_strings(p, a, EXTRA) }
    | a[asdl_expr_seq*]=tstring+ { _PyPegen_concatenate_tstrings(p, a, EXTRA) }

...

invalid_string_tstring_concat:
    | (fstring|string)+ a[expr_ty]=tstring { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "cannot mix t-strings with strings or f-strings") }
    | tstring+ a[expr_ty]=(fstring|string) { RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "cannot mix t-strings with strings or f-strings") }
```

or do we prefer separating `strings` from `tstrings` (and I guess having two different `invalid` rules)

lysnikolaou replied:
Sure, this looks even better. Can we use `RAISE_SYNTAX_ERROR_KNOWN_RANGE` instead of `KNOWN_LOCATION` so that the error is a bit clearer?

lysnikolaou replied:
Also, is this ready for another review or should I wait for a ping from you?

davepeck replied:
> Can we use RAISE_SYNTAX_ERROR_KNOWN_RANGE

Ah, yeah, that'd be nicer; will do. I guess the "ideal" range would include exactly two neighboring literals that conflict.

> Also, is this ready for another review or should I wait for a ping from you?

Not yet, I'll ping you. Thanks! (Still have more to tease apart in `action_helpers.c` and also need to tackle `ast_unparse`.)

sobolevn reviewed (commented):
Thank you! Several small nitpicks :)

lysnikolaou reviewed (changes requested):
Thanks @davepeck for working on this! Looks great in general. Left a few unimportant comments and one more significant one about how to implement this in the parser. Also, we'll need to change `ast_unparse.c` around https://github.com/python/cpython/blob/main/Python/ast_unparse.c#L715.

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Two minor tweaks

davepeck committed:
Use subtests; test both directions.

davepeck committed:
Update Objects/templateobject.c

Co-authored-by: sobolevn <mail@sobolevn.me>

davepeck committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2025-07-08-23-22-08.gh-issue-132661.34ftJl.rst

Co-authored-by: sobolevn <mail@sobolevn.me>

davepeck committed:
Fix bug in suggested change

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
First pass at update to grammar and action_helpers.c

Squashed commit of the following:

commit 0941a3088578c1a8002e58ea6e11527bedf0e81e
Author: Dave <davepeck@gmail.com>
Date:   Thu Jul 10 01:49:24 2025 +0000

    Maybe cleaner?

commit e5a69c851a0f43a4e84325051a37de20282dd810
Author: Dave <davepeck@gmail.com>
Date:   Wed Jul 9 23:17:58 2025 +0000

    In progress: update grammar

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Simplify grammar

davepeck committed:
Use KNOWN_RANGE for string/t-string mix error messages

davepeck committed:
Merge branch 'pep750-concat-update' of github.com:t-strings/cpython into pep750-concat-update

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Clean up ast_unparse.c append_tstring()

lysnikolaou commented on Grammar/python.gram:
```diff
@@ -1553,6 +1556,12 @@ invalid_tstring_conversion_character:
 
+invalid_string_tstring_concat:
+    | a=(fstring|string)+ b[expr_ty]=tstring {
+        RAISE_SYNTAX_ERROR_KNOWN_RANGE(PyPegen_last_item(a, expr_ty), b, "cannot mix t-strings with strings or f-strings") }
```
Maybe we can change the error message here to something like `cannot mix t-string literals with string or byte literals` to cover all of the cases.

Ditto below.

lysnikolaou commented on Misc/NEWS.d/next/Core_and_Builtins/2025-07-08-23-22-08.gh-issue-132661.34ftJl.rst:
```diff
@@ -0,0 +1,4 @@
+Reflect recent :pep:`750` change.
+
+Disallow concatenation of :class:`string.Template` and :class:`str`. Also, disallow implicit
```
```suggestion
Disallow concatenation of :class:`string.templatelib.Template` and :class:`str`. Also, disallow implicit
```

davepeck replied:
I removed the `:class:` prefix since there's no documentation -- probably good enough and we can go back and edit elsewhere.

lysnikolaou commented on Objects/templateobject.c:
```diff
@@ -344,47 +297,17 @@ template_concat_templates(templateobject *self, templateobject *other)
+
+    PyErr_Format(PyExc_TypeError,
+        "can only concatenate Template (not \"%T\") to Template",
+        other);
```
Let's include the full path to `Template` since just `Template` is ambiguous.

```suggestion
    PyErr_Format(PyExc_TypeError,
        "can only concatenate string.templatelib.Template (not \"%T\") to string.templatelib.Template",
        other);
```

lysnikolaou commented on Parser/action_helpers.c:
```diff
@@ -1872,23 +1871,20 @@ _PyPegen_concatenate_strings(Parser *p, asdl_expr_seq *strings,
                 break;
-            case TemplateStr_kind:
-                t_string_found = 1;
-                break;
```
Maybe leave this in and do an `assert(0)` with a comment that the grammar handles this?

davepeck replied:
Ah, I like that. Done!

lysnikolaou commented on Python/ast_unparse.c:
```diff
@@ -700,62 +700,11 @@ append_templatestr(PyUnicodeWriter *writer, expr_ty e)
-
-error:
+    int result = _write_values_subarray(writer, e->v.TemplateStr.values,
+            0, len - 1, 't', arena);
```
I'm pretty sure we can get rid of `_write_values_subarray` and just call `build_ftstring_body` directly with `e->v.TemplateStr.values`. That way, we also don't need an arena, which is always nice.

davepeck replied:
Whoops, I totally missed this. Fixed now; thanks!

lysnikolaou reviewed (commented):
Looks even better this time. Thanks @davepeck! Left a few comments inline again. It'll probably be good to go once these are addressed.

Regarding your questions:

> 1. I suppose we _could_ break `_build_concatenated_str()` in `action_helpers.c` into two separate routines. But given how much duplicated logic there would be around flattening/concatenating/re-folding, I decided to leave it untouched. Let me know if you'd like me to tease this apart further.

Probably okay to leave `_build_concatenated_str` as-is. There would be too much duplication for us to break it into two separate functions, and even adding more asserts would make things more complicated IMO.

> 2. Because t-string/non-t-string concat errors are handled in the grammar, but bytes/nonbytes are handled in `action_helpers.c`, `t"foo" b"bar"` results in the syntax error `"cannot mix t-strings with strings or f-strings"`. Are we okay with that?

Addressed inline.

davepeck committed:
Merge branch 'main' into pep750-concat-update

davepeck committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2025-07-08-23-22-08.gh-issue-132661.34ftJl.rst

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

davepeck committed:
Update Objects/templateobject.c

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

davepeck committed:
Update error messages; add an assert(0)

davepeck committed:
Simplify ast_unparase.c further

davepeck committed:
Fix news dangling reference (for now).

davepeck committed:
Merge branch 'pep750-concat-update' of github.com:t-strings/cpython into pep750-concat-update

efimov-mikhail commented on Lib/test/test_tstring.py:
```diff
@@ -161,9 +160,9 @@ def test_template_concatenation(self):
-        self.assertEqual(fstring(combined), "Hello, world")
+        expected_msg = 'can only concatenate Template (not "str") to Template'
+        with self.assertRaises(TypeError, msg=expected_msg):
+            t1 + ", world"
```
```suggestion
        expected_msg = 'can only concatenate string.templatelib.Template ' \
            '\\(not "str"\\) to string.templatelib.Template'
        with self.assertRaisesRegex(TypeError, expected_msg):
            t1 + ", world"
```

efimov-mikhail replied:
Current `self.assertRaises` call just checks presence of `TypeError`, not its message.

davepeck replied:
Oh, good catch.

efimov-mikhail commented on Lib/test/test_tstring.py:
```diff
@@ -174,9 +173,9 @@ def test_template_concatenation(self):
-        self.assertEqual(fstring(t), "Hello, Python")
+        expected_msg = 'can only concatenate str (not "string.templatelib.Template") to str'
+        with self.assertRaises(TypeError, msg=expected_msg):
+            "Hello, " + t"{name}"
```
```suggestion
        expected_msg = 'can only concatenate str ' \
            '\\(not "string.templatelib.Template"\\) to str'
        with self.assertRaisesRegex(TypeError, expected_msg):
            "Hello, " + t"{name}"
```

davepeck committed:
Fix exception assert message checks

davepeck committed:
Remove no-longer used path in codegen.c

lysnikolaou reviewed (approved):
Looks good to me! Thanks @davepeck! Great work.
